flight/Modules/Actuator/actuator.c

```diff
@@ -100,24 +101,102 @@ static float motor_mixer[MAX_MIX_ACTUATORS * MIXERSETTINGS_MIXER1VECTOR_NUMELEM]
 static ActuatorSettingsData actuatorSettings;
 static SystemSettingsAirframeTypeOptions airframe_type;
 
-static float curve1[MIXERSETTINGS_THROTTLECURVE1_NUMELEM];
 static float curve2[MIXERSETTINGS_THROTTLECURVE2_NUMELEM];
 
 static MixerSettingsCurve2SourceOptions curve2_src;
 
 // Private functions
 static void actuator_task(void* parameters);
 
-static float scale_channel(float value, int idx);
+static float scale_channel(float value, int idx, bool active_cmd);
 static void set_failsafe();
 
-static float throt_curve(const float input, const float *curve,
-		uint8_t num_points);
 static float collective_curve(const float input, const float *curve,
 		uint8_t num_points);
 
 volatile enum actuator_interlock actuator_interlock = ACTUATOR_INTERLOCK_OK;
 
+struct dshot_command {
+	union {
+		uint32_t value;
+		struct {
+			uint8_t cmd_id;
+			uint8_t num_to_send;
+			uint16_t channel_mask;
+		};
+	};
+};
+
+static volatile struct dshot_command pending_cmd;
+static struct dshot_command cur_cmd;
+
+static uint16_t desired_3d_mask;
+
+#define DSHOT_COMMAND_DONTSPIN 0
+#define DSHOT_COMMAND_NO3DMODE 9
+#define DSHOT_COMMAND_3DMODE   10
+
+#define DSHOT_COUNT_EXCESSIVE 12
+
+static bool actuator_get_dshot_command(bool armed) {
+	/* Finish off any current command. */
+	if (cur_cmd.num_to_send) {
+		cur_cmd.num_to_send--;
+		return true;
+	}
+
+	if (armed) {
+		return false;
+	}
+
+	uint32_t value = pending_cmd.value;
+
+	if (!value) {
+		return false;
+	}
+
+	pending_cmd.value = 0;
+	cur_cmd.value = value;
+
+	if (cur_cmd.num_to_send) {
+		cur_cmd.num_to_send--;
+		return true;
+	}
+
+	return false;
+}
+
+int actuator_send_dshot_command(uint8_t cmd_id, uint8_t num_to_send,
+		uint16_t channel_mask)
+{
+	struct dshot_command new_cmd;
+
+	new_cmd.cmd_id = cmd_id;
+	new_cmd.num_to_send = num_to_send;
+	new_cmd.channel_mask = channel_mask;
+
+	if (__sync_bool_compare_and_swap(&pending_cmd.value, 0,
```

Comment:
This is an atomic in preparation for other modules to trigger dshot commands when disarmed (e.g. actuator beep)